src/vs/editor/contrib/stickyScroll/browser/stickyScrollModelProvider.ts

```diff
@@ -0,0 +1,411 @@
+
+	/**
+	 * Method which computes the sticky model and returns a status to signal whether the sticky model has been successfully found
+	 * @param textmodel text-model of the editor
+	 * @param modelVersionId version ID of the text-model
+	 * @param token cancellation token
+	 * @returns a promise of a status indicating whether the sticky model has been successfully found
+	 */
+	computeStickyModel(textmodel: ITextModel, modelVersionId: number, token: CancellationToken): Promise<Status> | Status;
+}
+
+abstract class StickyModelCandidateProvider implements IStickyModelCandidateProvider {
+
+	private _providerModelPromise: CancelablePromise<any> | null = null;
+	protected _stickyModel: StickyModel | null = null;
+
+	constructor() { }
+
+	get stickyModel(): StickyModel | null {
+		return this._stickyModel;
+	}
+
+	get providerModelPromise(): CancelablePromise<any> | null {
+		return this._providerModelPromise;
+	}
+
+	private _invalid(): Status {
+		this._stickyModel = null;
+		return Status.INVALID;
+	}
+
+	public computeStickyModel(textModel: ITextModel, modelVersionId: number, token: CancellationToken): Promise<Status> | Status {
+		if (!this.isProviderValid(textModel)) {
+			return this._invalid();
+		}
+		const providerModelPromise = this._providerModelPromise = createCancelablePromise(token => this.createModelFromProvider(textModel, modelVersionId, token));
+
+		return providerModelPromise.then(providerModel => {
+			if (!this.isModelValid(providerModel)) {
+				return this._invalid();
+
+			}
+			if (token.isCancellationRequested) {
+				return Status.CANCELED;
+			}
+			this._stickyModel = this.createStickyModel(textModel, modelVersionId, token, providerModel);
+			return Status.VALID;
+		}).then(undefined, (err) => {
+			onUnexpectedError(err);
+			return Status.CANCELED;
+		});
+	}
+
+	/**
+	 * Method which checks whether the model returned by the provider is valid and can be used to compute a sticky model.
+	 * This method by default returns true.
+	 * @param model model returned by the provider
+	 * @returns boolean indicating whether the model is valid
+	 */
+	protected isModelValid(model: any): boolean {
+		return true;
+	}
+
+	/**
+	 * Method which checks whether the provider is valid before applying it to find the provider model.
+	 * This method by default returns true.
+	 * @param textModel text-model of the editor
+	 * @returns boolean indicating whether the provider is valid
+	 */
+	protected isProviderValid(textModel: ITextModel): boolean {
+		return true;
+	}
+
+	/**
+	 * Abstract method which creates the model from the provider and returns the provider model
+	 * @param textModel	text-model of the editor
+	 * @param textModelVersionId text-model version ID
+	 * @param token cancellation token
+	 * @returns the model returned by the provider
+	 */
+	protected abstract createModelFromProvider(textModel: ITextModel, textModelVersionId: number, token: CancellationToken): any;
```

Comment:
The `any` as result is not nice

Reply:
Make this a type variable of the StickyModelCandidateProvider:
- StickyModelCandidateProvider<T>
- StickyModelCandidateProvider: Promise<T>